forge/src/main/java/org/openjdk/skara/forge/github/GitHubRepository.java

```diff
@@ -293,9 +293,14 @@ public long id() {
     }
 
     @Override
-    public Hash branchHash(String ref) {
-        var branch = request.get("branches/" + ref).execute();
-        return new Hash(branch.get("commit").get("sha").asString());
+    public Optional<Hash> branchHash(String ref) {
+        var branch = request.get("branches/" + ref)
+                .onError(r -> r.statusCode() == 404 ? Optional.of(JSON.object().put("NOT_FOUND", true)) : Optional.empty())
+                .execute();
+        if (branch.contains("NOT_FOUND")) {
```

Comment:
I'm trying to figure out the logic here. So for a 404, you return a special "NOT_FOUND" object, which you detect after the request is done, to return Optional.empty().

But for any other errors..? This `var` syntax might be convenient when working in an IDE, but it sure makes code reviewing difficult. I assume `branch` is a JSON object, and that the results from onError are unwrapped before returning it from the request. But how will this work with the return statement below? 

branch.get("commit").get("sha") seems like it will fail with an exception...

Reply:
This is an established pattern in Skara that I Just copied. The `onError` lambda returns an optional. If that optional is empty, then the default error behavior happens, which is to throw exception. So by only returning an object on 404, then that's the only result code we are changing the behavior on.

Reply:
Aha, I see. A bit odd design, but it works I guess.